src/protocols/secure_channel/CASESession.cpp

```diff
@@ -404,187 +445,283 @@ CHIP_ERROR CASESession::HandleSigmaR1(System::PacketBufferHandle && msg)
     ChipLogDetail(SecureChannel, "Peer assigned session key ID %d", initiatorSessionId);
     SetPeerSessionId(initiatorSessionId);
 
-    SuccessOrExit(err = tlvReader.Next());
-    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
-    SuccessOrExit(err = tlvReader.GetBytes(destinationIdentifier, sizeof(destinationIdentifier)));
+    if (sessionResumptionRequested && memcmp(resumptionID, mResumptionId, kCASEResumptionIDSize) == 0)
+    {
+        // Cross check resume1MIC with the shared secret
+        SuccessOrExit(err = ValidateSigmaResumeMIC(resume1MICSpan, ByteSpan(initiatorRandom), ByteSpan(mResumptionId),
+                                                   ByteSpan(kKDFS1RKeyInfo), ByteSpan(kResume1MIC_Nonce)));
 
+        // Send Sigma2Resume message to the initiator
+        SuccessOrExit(err = SendSigma2Resume(ByteSpan(initiatorRandom)));
+    }
+    else
     {
         const ByteSpan * ipkListSpan = GetIPKList();
         FabricIndex fabricIndex      = Transport::kUndefinedFabricIndex;
         memcpy(mIPK, ipkListSpan->data(), sizeof(mIPK));
+
+        SuccessOrExit(err = tlvReader.Next());
+        VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+        SuccessOrExit(err = tlvReader.GetBytes(destinationIdentifier, sizeof(destinationIdentifier)));
+
         VerifyOrExit(mFabricsTable != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
         fabricIndex = mFabricsTable->FindDestinationIDCandidate(ByteSpan(destinationIdentifier), ByteSpan(initiatorRandom),
                                                                 ipkListSpan, GetIPKListEntries());
         VerifyOrExit(fabricIndex != Transport::kUndefinedFabricIndex, err = CHIP_ERROR_CERT_NOT_TRUSTED);
 
         mFabricInfo = mFabricsTable->FindFabricWithIndex(fabricIndex);
         VerifyOrExit(mFabricInfo != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
-    }
 
-    SuccessOrExit(err = tlvReader.Next());
-    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
-    SuccessOrExit(err = tlvReader.GetBytes(mRemotePubKey, static_cast<uint32_t>(mRemotePubKey.Length())));
+        SuccessOrExit(err = tlvReader.Next());
+        VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+        SuccessOrExit(err = tlvReader.GetBytes(mRemotePubKey, static_cast<uint32_t>(mRemotePubKey.Length())));
+
+        SuccessOrExit(err = SendSigma2());
+    }
 
 exit:
 
-    if (err != CHIP_NO_ERROR)
+    if (err == CHIP_ERROR_CERT_NOT_TRUSTED)
+    {
+        SendStatusReport(mExchangeCtxt, kProtocolCodeNoSharedRoot);
```

Comment:
kProtocolCodeNoSharedRoot is only used when no DestinationID candidate matches. We should change DestinationId matching:

```
        fabricIndex = mFabricsTable->FindDestinationIDCandidate(ByteSpan(destinationIdentifier), ByteSpan(initiatorRandom),
                                                                ipkListSpan, GetIPKListEntries());
        VerifyOrExit(fabricIndex != Transport::kUndefinedFabricIndex, err = CHIP_ERROR_CERT_NOT_TRUSTED);
```

to use `CHIP_ERROR_KEY_NOT_FOUND` instead, since there are no certs checked in this step... It's a minor rename, but one that makes the code clearer.